h2o-core/src/main/java/hex/ModelMetricsRegressionCoxPH.java

```diff
@@ -147,83 +134,455 @@ long discordant() {
+
+      Frame frame = new MRTask() {
+        @Override
+        public void map(Chunk c0, Chunk c1, NewChunk nc) {
+          for (int i = 0; i < c0._len; i++)
+            nc.addNum(c1.atd(i) - c0.atd(i));
+        }
+      }.doAll(1, Vec.T_NUM, fr1)
+              .outputFrame(new String[]{"durations"}, null);
+      
+      return frame.vec(0);
+    }
+
+    private static Stats concordanceStats(Frame fr){
+      final Frame withoutNas = removeNAs(fr);
+
+      final int[] stratasAndDuration = new int[withoutNas.numCols() - 2];
+        final int[] strataIndexes = new int[withoutNas.numCols() - 3];
+        for (int i = 0; i < strataIndexes.length; i++) {
+          stratasAndDuration[i] = i + 3;
+          strataIndexes[i] = i + 3;
+        }
+        stratasAndDuration[withoutNas.numCols() - 3] = 0;
+
+        if (0 == withoutNas.numRows()) {
+          return new Stats();
+        }
+
+        final Frame sorted = withoutNas.sort(stratasAndDuration);
+
+        final List<Vec.Reader> strataCols = stream(strataIndexes).boxed().map(i -> sorted.vec(i).new Reader()).collect(toList());
+
+        long lastStart = 0L;
+        List lastRow = new ArrayList(sorted.numCols() - 3);
+        Stats statsAcc = new Stats();
+
+        for (long i = 0; i < sorted.numRows(); i++) {
+          final List row = new ArrayList(sorted.numCols() - 3);
+          for (Vec.Reader strataCol : strataCols) {
+            row.add(strataCol.at(i));
+          }
+
+          if (!lastRow.equals(row)) {
+            lastRow = row;
+            Stats stats = statsForAStrata(sorted.vec("duration").new Reader()
+                    , sorted.vec("event").new Reader()
+                    , sorted.vec("estimate").new Reader()
+                    , lastStart
+                    , i);
+            lastStart = i;
+
+            statsAcc = statsAcc.plus(stats);
+          }
+        }
+
+      Stats stats = statsForAStrata( sorted.vec("duration").new Reader()
+                                   , sorted.vec("event").new Reader()
+                                   , sorted.vec("estimate").new Reader()
+                                   , lastStart
+                                   , sorted.numRows());
+
 
-      final Stats stats = concordanceStats(null == startVec ? null : startVec.new Reader(), 
-              stopVec.new Reader(), eventVec.new Reader(),
-              strataVecs.stream().map(it -> it.new Reader()).collect(toList()),
-              estimateVec.new Reader(), length);
+      return statsAcc.plus(stats);
+    }
+
+    private static Frame removeNAs(Frame fr) {
+      final int[] iDontWantNAsInThisCols = new int[]{0, 2};
+
+      final Frame withoutNas = new Merge.RemoveNAsTask(iDontWantNAsInThisCols)
+                                        .doAll(fr.types(), fr)
+                                        .outputFrame(fr.names(), fr.domains());
 
-      return stats;
+      withoutNas.replace(1, withoutNas.vec("event").toNumericVec());
```

Comment:
removeNAs is also converting event column to numeric? 

I would expect such transformations to be done in top-level (public) methods instead of utility functions.

Reply:
was removed